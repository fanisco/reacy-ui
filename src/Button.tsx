import React from 'react';
import styled from 'styled-components';

interface IProps {
    onClick: Function
}


export const Button: React.FC<IProps> = (props: IProps) => {
    const Button = styled.button``;
    return (
        <Button onClick={() => props.onClick()}>Button</Button>
    );
};
