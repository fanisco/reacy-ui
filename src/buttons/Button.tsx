import React from 'react';
import styled from 'styled-components';
import {Buttons, Core} from '../core/interfaces';
import {modsFromArray} from '../core/helpers';
import {button} from '../core/mixins/button';

export const Button: React.FC<Buttons.IButton> = ({onClick, href, children, mods}) => {
  if (href) {
    return <_Button as="a" onClick={onClick} href={href} mods={modsFromArray(mods)}>{children}</_Button>;
  } else {
    return <_Button onClick={onClick} mods={modsFromArray(mods)}>{children}</_Button>;
  }
};

const _Button = styled.button<Core.IStyled>`
  display: inline-flex;
  position: relative;
  box-sizing: border-box;
  font-weight: 400;
  cursor: pointer;
  text-align: left;
  text-decoration: none;
  border: 1px solid;
  transition: all 0.15s ease-in-out;
  white-space: nowrap;

  :hover { transition: all 0.15s ease-in-out; }
  :focus { outline: 0 none; }

  ${props => button(props.mods)}
`;
