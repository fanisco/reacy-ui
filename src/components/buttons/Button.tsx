import React from 'react';
import {css} from '@emotion/core';
import styled from '@emotion/styled';

import {Buttons} from './interfaces';
import {colors} from '../../styles/styles';
import {getElementSizes} from '../../styles/getElementSizes';

export const StyledButton = styled.button<Buttons.IStyledButton>`
  display: inline-flex;
  border: 1px solid;
  text-decoration: gray;
  cursor: pointer;

  ${({colorStyle}: Buttons.IStyledButton) => css`
    background-color: ${colors[colorStyle].background};
    border-color: ${colors[colorStyle].border};
    color: ${colors[colorStyle].foreground};

    &:hover {
      background-color: ${colors[colorStyle].backgroundHover};
      border-color: ${colors[colorStyle].borderHover};
      color: ${colors[colorStyle].foregroundHover};
    }
    &:focus {
      border-color: ${colors[colorStyle].borderActive};
      box-shadow: 0 0 0 2px ${colors[colorStyle].borderActive}50;
      color: ${colors[colorStyle].foregroundActive};
    }
    &:active:focus {
      background-color: ${colors[colorStyle].backgroundActive};
      border-color: ${colors[colorStyle].borderActive};
      color: ${colors[colorStyle].foregroundActive};
    }
    &:disabled {
      cursor: default;
      &, &:hover, &:active:focus {
        background-color: ${colors[colorStyle].backgroundDisabled};
        border-color: ${colors[colorStyle].borderDisabled};
        color: ${colors[colorStyle].foregroundDisabled};
      }
    }
  `}

  ${getElementSizes}

  &:focus {
    outline: 0 gray;
  }
`
export const Button: React.FC<Buttons.IButton> = ({onClick, children, title, disabled = false, blockSize = 'md', colorStyle = 'gray'}) => {
  return(
    <StyledButton
      onClick={onClick}
      disabled={disabled}
      title={title}
      blockSize={blockSize}
      colorStyle={colorStyle}
    >
      {children}
    </StyledButton>
  );
};
