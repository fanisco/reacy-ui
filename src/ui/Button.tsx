import React from 'react';
import styled from 'styled-components';
import { Colors, Fonts, Dims } from './constants';
import { Styles, Sizes } from './enums';

interface IProps {
    onClick?: Function;
    disabled?: boolean;
    style?: Styles;
    size?: Sizes;
}

export const Button: React.FC<IProps> = ({ style = Styles.default, size = Sizes.md, onClick, disabled, children }) => {
    const colors = Colors[style];
    const sizes = Dims[size];
    const Button = styled.button`
        height: ${sizes.elementHeight};
        padding: 0 ${sizes.elementPadding};
        background: ${colors.baseL1};
        border: 1px solid ${colors.baseL3};
        border-radius: ${sizes.borderRadius};
        color: ${colors.text};
        cursor: pointer;
        font: ${sizes.font}/1 "${Fonts.ff}";
        font-weight: 600;
        
        &:hover {
            background: ${colors.baseL2};
        }
        &:focus {
            outline: 0 none;
        }
        &:active:focus {
            background: ${colors.baseL3};
        }
        &:disabled {
            color: ${colors.baseL3};
            cursor: default;
        }
        &:disabled:hover {
            background: ${colors.baseL1}
        }
    `;
    return (
        <Button onClick={() => onClick && onClick()} disabled={disabled}>{children}</Button>
    );
};
