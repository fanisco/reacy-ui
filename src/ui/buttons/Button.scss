@import '../base/mixins';
.btn {
	$class: btn;
	display: flex;
	justify-content: space-between;
	align-items: center;
	position: relative;
	box-sizing: border-box;
	font-weight: 400;
	cursor: pointer;
	text-align: left;
	border: 1px solid;
	transition: all 0.15s ease-in-out;

	&:hover { transition: all 0.15s ease-in-out 0.15s; }
	&:focus { outline: 0 none; }

	@include btn-sizes(3, $class);
	@include btn-styles(1, $class);

	&_xs { @include btn-sizes(1, $class); }
	&_sm { @include btn-sizes(2, $class); }
	&_lg { @include btn-sizes(4, $class); }
	&_xl { @include btn-sizes(5, $class); }

	&_primary { @include btn-styles(2, $class); }
	&_secondary { @include btn-styles(3, $class); }
	&_success { @include btn-styles(4, $class); }
	&_warning { @include btn-styles(5, $class); }

	&_bold { font-weight: bold; }
	&_italic { font-style: italic; }
	&_rounded { border-radius: 50px; }
	&_fullWidth { width: 100%; }

	&_link {
		height: auto;
		padding: 0;
		background: none;
		border: 0 none;

		&:hover {
			background: none;
			text-decoration: underline;
		}
		&:active:focus {
			background: none;
			box-shadow: none;
		}
	}
	&_outline {
		background: none;
	}

	&:disabled:not(&_link), &:disabled:hover:not(&_link), &:disabled:active:focus:not(&_link) {
		background: none;
		background: $b250;
		border-color: $b250;
		color: $b100;
		cursor: default;
	}

	&_first {
		margin-right: -1px;
		border-top-right-radius: 0;
		border-bottom-right-radius: 0;
	}
	&_last {
		margin-right: -1px;
		border-top-left-radius: 0;
		border-bottom-left-radius: 0;
	}
	&_mid {
		margin-right: -1px;
		border-radius: 0;
	}
	&_first, &_last, &_mid {
		z-index: 1;
	}
	&_active#{&}_first, &_active#{&}_last, &_active#{&}_mid {
		z-index: 1;
	}

	&__icon {
		&_start {
			margin-right: 5px;
		}
		&_end {
			margin-left: 5px;
		}
	}
}
