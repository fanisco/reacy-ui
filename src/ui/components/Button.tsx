import React from 'react';
import styled from 'styled-components';
import { Colors, Dims, Fonts } from '../constants';
import { Sizes, Styles } from '../enums';

interface IProps {
    onClick?: Function;
    disabled?: boolean;
    style?: Styles;
    size?: Sizes;
    mode?: 'default' | 'link' | 'outline';
    textAlign?: 'left' | 'center' | 'right';
    fullWidth?: boolean;
}

export const Button: React.FC<IProps> =
    ({ style = Styles.default, size = Sizes.md, mode = 'default', textAlign = 'center', ...props }) => {
    const colors = getColors(style, Colors[style]);
    const dims = Dims[size];
    return (
        <ButtonElement
            dims={dims}
            colors={colors}
            mode={mode}
            textAlign={textAlign}
            fullWidth={props.fullWidth}
            onClick={() => props.onClick && props.onClick()}
            disabled={props.disabled}
        >{props.children}</ButtonElement>
    );
};

interface StyledProps {
    dims?: any;
    colors?: any;
    mode?: string;
    textAlign?: string;
    fullWidth?: boolean;
}

const ButtonElement = styled.button<StyledProps>`
    box-sizing: border-box;
    height: ${props => props.dims.elementHeight}px;
    padding: ${props => props.dims.spacings}px ${props => props.dims.spacing}px;
    background: ${props => props.colors.idleColor} linear-gradient(to bottom, rgba(255, 255, 255, ${props => props.colors.gradient}), rgba(255, 255, 255, 0));
    border: 1px solid ${props => props.colors.borderColor};
    border-radius: ${props => props.dims.borderRadius}px;
    color: ${props => props.colors.textColor};
    text-align: ${props => props.textAlign};
    font: ${props => props.dims.fontSize}px/${props => props.dims.lineHeight} "${Fonts.ff}";
    font-weight: 500;
    cursor: pointer;
    
    &:hover {
        border-color: ${props => props.colors.borderHoverColor};
        background-color: ${props => props.colors.hoverColor};
    }
    &:focus {
        outline: 0 none;
    }
    &:active:focus {
        background: ${props => props.colors.activeColor};
        border-color: ${props => props.colors.borderColor};
        box-shadow: inset 0 0 10px ${props => props.colors.shadowColor}50;
    }
    
    ${props => props.fullWidth && `
        width: 100%;
    `}
    
    ${props => props.mode === 'link' ? `
        height: auto;
        padding: 0;
        background: none;
        border: 0 none;
        font-weight: 400;
        color: ${props.colors.idleColor};

        &:hover {
            background: none;
            color: ${props.colors.hoverColor};
            text-decoration: underline;
        }
        &:active:focus {
            background: none;
            box-shadow: none;
            color: ${props.colors.activeColor};
        }
    ` : props.mode === 'outline' && `
        background: none;
        border-radius: 50px;
        color: ${props.colors.idleColor};

        &:hover {
            background: ${props.colors.hoverColor};
            border-color: ${props.colors.borderHoverColor};
            color: ${props.colors.textColor};
        }
        &:active:focus {
            background: ${props.colors.activeColor};
            border-color: ${props.colors.borderColor};
            color: ${props.colors.textColor};
            }
    `}
    
    &:disabled, &:disabled:hover, &:disabled:active:focus {
        background: none;
        border-color: ${Colors[Styles.default].baseL0};
        color: ${Colors[Styles.default].baseL0};
        cursor: default;
    }
`;

const getColors = (style: string, colors: any): Scheme => {
    if (style !== 'default') {
        return {
            textColor: colors.text,
            idleColor: colors.baseL2,
            hoverColor: colors.baseL3,
            activeColor: colors.baseL2,
            borderColor: colors.baseL0,
            borderHoverColor: colors.baseL2,
            shadowColor: colors.baseL0,
            gradient: 0.15
        };
    } else {
        return {
            textColor: colors.text,
            idleColor: colors.baseL2,
            hoverColor: colors.baseL3,
            activeColor: colors.baseL2,
            borderColor: colors.baseL0,
            borderHoverColor: colors.baseL0,
            shadowColor: colors.baseL0,
            gradient: 0.75
        };
    }
};

type Scheme = {
    textColor: string;
    idleColor: string;
    hoverColor: string;
    activeColor: string;
    borderColor: string;
    borderHoverColor: string;
    shadowColor: string;
    gradient: number;
}
