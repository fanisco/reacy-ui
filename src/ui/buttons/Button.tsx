import React from 'react';
import { Sizes, Styles } from '../enums';
import { Icon } from '../icons';
import './Button.scss';

export const Button: React.FC<IProps> = (props) => {
    return (
        <button
            className={getClass(props)}
            disabled={props.state === false}
            onClick={() => props.onClick && props.onClick()}
        >{props.icon ? <Icon name={props.icon}/> : null}{props.children}</button>
    );
};

interface IProps {
    onClick?: Function;
    style?: Styles;
    size?: Sizes;
    mode?: 'default' | 'link' | 'outline';
    textAlign?: 'left' | 'center' | 'right';
    bold?: boolean;
    italic?: boolean;
    rounded?: boolean;
    fullWidth?: boolean;
    state?: boolean | null;
    side?: boolean | null;
    icon?: string;
}

const getClass = (props: IProps) => {
    const base = 'btn';
    const classes: string[] = [base];
    if (props.style !== 'default') {
        classes.push(`${base}_${props.style}`);
    }
    if (props.size !== 'md') {
        classes.push(`${base}_${props.size}`);
    }
    if (props.state !== null) {
        classes.push(`${base}_${props.state ? 'active' : 'disabled'}`);
    }
    if (props.bold) {
        classes.push(`${base}_bold`);
    }
    if (props.italic) {
        classes.push(`${base}_italic`);
    }
    if (props.rounded) {
        classes.push(`${base}_rounded`);
    }
    if (props.fullWidth) {
        classes.push(`${base}_fullWidth`);
    }
    if (props.mode !== 'default') {
        classes.push(`${base}_${props.mode}`);
    }
    if (props.side !== undefined) {
        classes.push(`${base}_${props.side ? 'first' : props.side === false ? 'last' : 'mid'}`);
    }
    return classes.join(' ');
};
