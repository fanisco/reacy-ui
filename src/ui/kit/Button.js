import React from 'react';
import {bemClasses} from '../core/bem';
import './Button.scss';

const Button = ({caption, onClick, href, children, className, mods}) => {
  const classNameFinal = bemClasses({
    base: 'button',
    mods,
    adds: [className]
  });
  if (href) {
    return <a className={classNameFinal} onClick={onClick} href={href}>{caption}{children}</a>;
  } else {
    return <button className={classNameFinal} onClick={onClick}>{caption}{children}</button>;
  }
}

export default Button;
