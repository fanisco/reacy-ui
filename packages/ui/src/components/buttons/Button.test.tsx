import React from 'react';
import {shallow, mount} from 'enzyme';

import {Button} from './Button';

describe('<Button>', () => {
  it('should render without problems', () => {
    const button = shallow(<Button>Button-example</Button>);
    expect(button).toMatchSnapshot();
  });
  it('should apply modifiers', () => {
    const button = mount(<Button mods={['primary']}/>);
    expect(button.getDOMNode().className).toContain('primary');
  });
  it('should call `onClick` callback after button was clicked', () => {
    const onClick = jest.fn();
    const button = mount(<Button onClick={onClick}/>);
          button.simulate('click');
    expect(onClick).toHaveBeenCalledTimes(1);
  });
});
