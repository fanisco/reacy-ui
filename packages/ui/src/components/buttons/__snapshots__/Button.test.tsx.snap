// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button> should render without problems 1`] = `ShallowWrapper {}`;
